File: ui/src/services/ssh.js
import {baseUrl} from "../state/index.js";



// http://127.0.0.1:8080/api/ssh/group/content
export const getSSHGroupContent = async ( groupId = 'root' ) => {

    let sid = window.localStorage.getItem('sid')

    let res = await fetch(baseUrl.value + 'api/ssh/group/content?groupId=' + groupId,{
        method:'GET',
        headers: {
            'sid':sid
        }
    })

    return await res.json()
}

//http://127.0.0.1:8080/api/ssh/group/save
export const saveSSHGroup = async (name = '',parentId = '')=>{
    let sid = window.localStorage.getItem('sid')

    let res = await fetch(baseUrl.value + 'api/ssh/group/save?' + `name=${name}` + '&' + `parentId=${parentId}`,{
        method:'PUT',
        headers: {
            'sid':sid,
        },
    })

    return await res.json()
}


// http://127.0.0.1:8080/api/ssh/info/save


/**
 * @param {object} info 
 * @param {string} info.groupId
 * @param {string} info.addrPort
 * @param {string} info.username
 * @param {string} info.password
 * @returns 
 */
export const saveSSHInfo = async (info)=>{
    let sid = window.localStorage.getItem('sid')

    let res = await fetch(baseUrl.value + 'api/ssh/info/save',{
        method:'PUT',
        headers: {
            'sid':sid,
            'Content-Type':'application/json'
        },
        body:JSON.stringify(info)
    })

    return await res.json()
}


//http://127.0.0.1:8080/api/ssh/group/tree
export const getSSHGroupTree = async (id = 'root') => {
    let sid = window.localStorage.getItem('sid')

    let res = await fetch(baseUrl.value + 'api/ssh/group/tree?' + `rootId=${id}`,{
        method:'GET',
        headers: {
            'sid':sid,
        },
    })

    return await res.json()
}


// http://127.0.0.1:8080/api/ssh/group/delete
export const deleteSSHGroup = async (groupId = null) => {

    if (groupId == null) return
    let sid = window.localStorage.getItem('sid')


    let res = await fetch(baseUrl.value + 'api/ssh/group/delete?' + `groupId=${groupId}`,{
        method:'DELETE',
        headers: {
            'sid':sid,
        },
    })

    return await res.json()

}



// http://127.0.0.1:8080/api/ssh/group/update
/**
 * @param {object} info
 * @param {string} info.id
 * @param {string} info.name
 * @param {string} info.parent
 * @returns
 */
export const updateSSHGroup = async (info) => {
    let sid = window.localStorage.getItem('sid')

    let res = await fetch(baseUrl.value + 'api/ssh/group/update',{
        method:'POST',
        headers: {
            'sid':sid,
            'Content-Type':'application/json'
        },
        body:JSON.stringify(info)
    })

    return await res.json()
}

File: apps/center/api/handler/ssh.go
package handler

import (
	"apps/center/api/param"
	"apps/center/api/response"
	"apps/center/api/vo"
	"apps/center/global"
	"apps/center/model"
	"apps/center/service"
	"apps/common/utils"
	"github.com/gin-gonic/gin"
	"net/http"
)

func SSHHandlerLoad(e *gin.RouterGroup) {

	sshRouters := e.Group("/ssh")

	sshRouters.GET("/group/content", GetSSHInfo)

	sshRouters.PUT("/info/save", SaveSSHInfo)

	sshRouters.PUT("/group/save", NewGroup)

	sshRouters.POST("/info/update", UpdateSSHInfo)

	sshRouters.DELETE("/info/delete", DeleteSSHInfo)

	sshRouters.POST("/group/update", UpdateGroupInfo)

	sshRouters.GET("/group/tree", GroupTree)

	sshRouters.DELETE("/group/delete", DeleteGroup)

}

// GetSSHInfo 获取组内信息（包括子组，和ssh信息）
// GET
// @Param groupId 组Id
// @Response []vo.GroupContent
func GetSSHInfo(c *gin.Context) {

	id, ok := c.GetQuery("groupId")

	if !ok {
		c.JSON(http.StatusOK, response.ParamsError())
		return
	}

	// response
	var groupContent []vo.GroupContent

	//检查 groupId 所对应的 ssh_group 是否存在

	// 检查是否存在对应 group
	tx := global.DB.Where("id = ?", id).Find(&model.SSHGroup{})
	rows := tx.RowsAffected
	if rows == 0 {
		c.JSON(http.StatusOK, response.StringFail("could not find ssh_group that id = "+id))
		return
	}

	var sshInfo []model.SSHInfo
	var sshGroup []model.SSHGroup

	global.DB.Where("group_id = ?", id).Find(&sshInfo)
	global.DB.Where("parent = ?", id).Find(&sshGroup)

	for _, group := range sshGroup {
		groupContent = append(groupContent, vo.GroupContent{
			Type:    vo.GROUP,
			Payload: group,
		})
	}

	for _, info := range sshInfo {
		info.Password = "******"
		groupContent = append(groupContent, vo.GroupContent{
			Type:    vo.INFO,
			Payload: info,
		})
	}

	c.JSON(http.StatusOK, response.Success(groupContent))
}

// SaveSSHInfo 保存ssh信息
// Post
// @Body param.SSHInfoParams
// @Response bool
func SaveSSHInfo(c *gin.Context) {

	params := param.SSHInfoParams{}
	err := c.ShouldBindJSON(&params)

	if err != nil {
		c.JSON(http.StatusOK, response.ParamsError())
		return
	}

	// 检查是否存在对应 group
	tx := global.DB.Where("id = ?", params.GroupId).Find(&model.SSHGroup{})
	rows := tx.RowsAffected
	if rows == 0 {
		c.JSON(http.StatusOK, response.StringFail("could not find ssh_group that id = "+params.GroupId))
		return
	}

	err = global.DB.Save(&model.SSHInfo{
		Id:       utils.UUID(),
		Username: params.Username,
		AddrPort: params.AddrPort,
		Password: params.Password,
		GroupId:  &params.GroupId,
	}).Error

	if err != nil {
		c.JSON(http.StatusOK, response.Fail(err))
		return
	}
	c.JSON(http.StatusOK, response.Success(true))
}

// NewGroup 创建组

func NewGroup(c *gin.Context) {

	name, n := c.GetQuery("name")
	parentId, p := c.GetQuery("parentId")

	if !(n && p) {
		c.JSON(http.StatusOK, response.ParamsError())
		return
	}

	// 检查是否存在对应 group
	tx := global.DB.Where("id = ?", parentId).Find(&model.SSHGroup{})
	rows := tx.RowsAffected
	if rows == 0 {
		c.JSON(http.StatusOK, response.StringFail("could not find ssh_group that id = "+parentId))
		return
	}

	group := model.SSHGroup{
		Id:     utils.UUID(),
		Name:   name,
		Parent: &parentId,
	}

	err := global.DB.Save(&group).Error

	if err != nil {
		c.JSON(http.StatusOK, response.Fail(err))
		return
	}

	c.JSON(http.StatusOK, response.Success(true))
}

// UpdateGroupInfo 更新组信息
// POST
// @Body model.SSHGroup
// @Response true or false
func UpdateGroupInfo(c *gin.Context) {

	var groupInfo model.SSHGroup

	err := c.ShouldBindJSON(&groupInfo)

	if err != nil {
		c.JSON(http.StatusOK, response.ParamsError())
		return
	}

	err = global.DB.Save(&groupInfo).Error

	if err != nil {
		c.JSON(http.StatusOK, response.Fail(err))
		return
	}

	c.JSON(http.StatusOK, response.Success(true))
}

func UpdateSSHInfo(c *gin.Context) {

	info := model.SSHInfo{}
	err := c.ShouldBindJSON(&info)

	if err != nil {
		c.JSON(http.StatusOK, response.ParamsError())
		return
	}

	// 检查是否存在对应 group
	tx := global.DB.Where("id = ?", *info.GroupId).Find(&model.SSHGroup{})
	rows := tx.RowsAffected
	if rows == 0 {
		c.JSON(http.StatusOK, response.StringFail("could not find ssh_group that id = "+*info.GroupId))
		return
	}

	err = global.DB.Updates(&info).Error

	if err != nil {
		c.JSON(http.StatusOK, response.Fail(err))
		return
	}

	c.JSON(http.StatusOK, response.Success(true))
}

func DeleteSSHInfo(c *gin.Context) {

	id, ok := c.GetQuery("infoId")

	if !ok {
		c.JSON(http.StatusOK, response.ParamsError())
		return
	}

	err := global.DB.Where("id = ?", id).Delete(&model.SSHInfo{}).Error

	if err != nil {
		c.JSON(http.StatusOK, response.Fail(err))
		return
	}

	c.JSON(http.StatusOK, response.Success(true))

}

func GroupTree(c *gin.Context) {

	rootId, ok := c.GetQuery("rootId")

	if !ok {
		c.JSON(http.StatusOK, response.ParamsError())
		return
	}

	var root model.SSHGroup

	r := service.SSHCRUD.Db.Where("id = ?", rootId).Find(&root).RowsAffected

	if r == 0 {
		c.JSON(http.StatusOK, response.StringFail("could not find ssh_group that id = "+rootId))
		return
	}

	tree := service.SSHCRUD.GroupTree(rootId)

	rootItem := vo.GroupTreeItem{
		Id:       root.Id,
		Name:     root.Name,
		Children: tree,
	}

	c.JSON(http.StatusOK, response.Success([]vo.GroupTreeItem{rootItem}))
}

// DeleteGroup 删除一个组（SSH）里的所有内容
// DELETE
// @Param groupId 组Id
// @Response bool 是否删除成功
func DeleteGroup(c *gin.Context) {

	groupId, ok := c.GetQuery("groupId")

	if !ok {
		c.JSON(http.StatusOK, response.ParamsError())
		return
	}

	err := service.SSHCRUD.DeleteGroup(groupId)

	if err != nil {
		c.JSON(http.StatusOK, response.Fail(err))
		return
	}

	c.JSON(http.StatusOK, response.Success(true))
}
